vulcan-core/scans/client.go

```diff
@@ -0,0 +1,185 @@
+	return fmt.Sprintf("unexpected http status, expected %d, got %d", e.Expected, e.Got)
+}
+
+// ScanData define the data of a scan retrieved by the scans client.
+type ScanData struct {
+	Date    string
+	Reports []report.Report
+}
+
+// ReportClient defines the services needed by the concurrent client.
+type ReportClient interface {
+	GetReport(date, scanID, checkID string) (*report.Report, error)
+}
+
+// Client allows to get the reports of a scan concurrently.
+type Client struct {
+	coreC   *coreclient.Client
+	report  ReportClient
+	workers int
+}
+
+// NewClient returns a new client scans client.
+func NewClient(coreClient *coreclient.Client, reportClient ReportClient, workers int) *Client {
+	return &Client{coreClient, reportClient, workers}
+}
+
+// Data downloads all the reports of a given scan using, at most, the
+// specified number of goroutines in the client.
+func (c *Client) Data(ctx context.Context, ID string) (ScanData, error) {
+	scanID, err := uuid.FromString(ID)
+	if err != nil {
+		return ScanData{}, err
+	}
+
+	resp, err := c.coreC.ShowScans(ctx, coreclient.ShowScansPath(scanID))
+	if err != nil {
+		return ScanData{}, err
+	}
+
+	if resp.StatusCode != http.StatusOK {
+		err = &ErrorUnexpectedStatus{http.StatusOK, resp.StatusCode}
+		return ScanData{}, fmt.Errorf("error querying scan in vulcan core %w", err)
+	}
+
+	scan, err := c.coreC.DecodeScan(resp)
+	if err != nil {
+		return ScanData{}, err
+	}
+
+	// We need the date of the scan in order to build the path for the reports
+	// of the checks belonging to it.
+
+	scanDate := scan.Scan.CreatedAt.Format("2006-01-02")
+
+	resp, err = c.coreC.ChecksScans(ctx, coreclient.ChecksScansPath(scanID))
+	if resp.StatusCode != http.StatusOK {
+		err = &ErrorUnexpectedStatus{http.StatusOK, resp.StatusCode}
+		return ScanData{}, fmt.Errorf("error querying scan checks in vulcan core %w", err)
+	}
+	checksData, err := c.coreC.DecodeChecks(resp)
+	if err != nil {
+		return ScanData{}, err
+	}
+	if checksData == nil {
+		return ScanData{}, errors.New("unexpected nil getting checks for a scan")
+	}
+
+	rs := &reportResults{}
+	rs.reports = make([]reportResult, 0, len(checksData.Checks))
+	wg := &sync.WaitGroup{}
+	checks := make(chan check, c.workers)
+
+	// Spawn a goroutine to spawn to the channel that the workers will read from.
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+	LOOP:
+		for _, c := range checksData.Checks {
+			select {
+			case <-ctx.Done():
+				break LOOP
+			default:
+				checks <- check{ID: c.ID.String(), ScanID: ID, Date: scanDate}
+			}
+		}
+		close(checks)
+	}()
+
+	// Spawn the workers.
+	wg.Add(1)
+	go func() {
+		cctx, _ := context.WithCancel(ctx)
+		defer wg.Done()
+		downloadCheckReport(wg, cctx.Done(), checks, rs, c.report)
+	}()
+	wg.Wait()
+	if ctx.Err() != nil {
+		// The download process has been canceled.
+		return ScanData{}, ctx.Err()
+	}
+	var result []report.Report
+	for _, r := range rs.reports {
+		if r.err != nil {
+			return ScanData{}, r.err
```

Comment:
In the current implementation, if we find a failed report we return an empty `ScanData{}`. 
Imagine a scenario where we have a scan that should contain 1000 checks, but when we query Results, it contains actually just 3 checks. This means that 3 requests will succeed while 997 will fail. Since we are returning an empty `ScanData{}` when we find the first error, can we modify the code to stop trying requests when it finds the first error, this way avoiding unnecessary requests?

Reply:
You are right here. The thing is that, indeed, what we need is to not return an error event if a check does not have a report, because we can have checks that can be failed because for instance the target does not exist. So what I'm going to do is to modify the client to return for each check, either an error or a report, and let caller decide what to do when the client was unable to get all the reports.